### PR TITLE
Sane buildflow

### PR DESCRIPTION
**Pull request by yne**
## Description

Based on existing work done by @ILOVEPIE in https://github.com/opentypejs/opentype.js/pull/693 this PR migrate to a sane build pipeline (which is urgent) while leaving any compatibility-related change to https://github.com/opentypejs/opentype.js/pull/693 (which is nice, but less urgent).

- Rename `src/*.js` to `src/*.mjs` and `test/*.js` to `test/*.spec.mjs`
- Remove unused reify dependency
- Use latest Mocha => no more critical vulnerabilities when `npm i`
- Use latest ESLint 9 => need to migrate to they new config format
- Deprecate load*/download functions (https://github.com/opentypejs/opentype.js/issues/675) and remove associated tests
- Remove dead code (private load*() function, isNode/isBrowser...)
- Fix some issues in recent tests addition
- Add support for many popular test runner:
```sh
npx mocha
npx bun test
npx jasmine "**/*.spec.mjs" # may fail on strict compare
```
*BONUS:* This PR also cleanup our ESLint rules:
- "ban forEach" is migrate to a "selector" rule
- prevent import loop => did nothing
- force import file extension => esbuild enforce it already

> [!TIP]
> This halves (243 => 144) our NPM dependencies

## Motivation and Context

Lower our attack surface by half, while keeping the backward-compatibility improvment in https://github.com/opentypejs/opentype.js/pull/693

## How Has This Been Tested?

```
npm i && npm run test
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] I did `npm run test` and all tests passed green (including code styling checks).
- [x] I have added tests to cover my changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the **README** accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: test/bidi.spec.mjs [hunk 1]
```diff
@@ -98,12 +98,12 @@ describe('bidi.js', function() {
...
 
-    describe('thai scripts', () => {
+    describe('thai scripts', function() {
 
         let thaiFont;
         let bidiThai;
 
-        before(()=> {
+        beforeEach(()=> {
             thaiFont = loadSync('./test/fonts/NotoSansThai-Medium-Testing-v1.ttf');
```
Comment (Connum) [inline test/bidi.spec.mjs:107]: Why do we need to load/parse the font before each test? If it's not being modified and therefore interfering with following tests, this just adds overhead and increases testing runtime.

File: test/bidi.spec.mjs [hunk 2]
```diff
@@ -133,24 +133,24 @@ describe('bidi.js', function() {
...
                 assert.deepEqual(words, ['เป็นคนใจดีสำหรับทุกคน', 'ความรักคือทุกสิ่ง']);
             });
         });
     });
 
-    describe('noto emoji with ccmp', () => {
+    describe('noto emoji with ccmp', function () {
         let notoEmojiFont;
-        before(()=> {
+        beforeEach(()=> {
```
Comment (Connum) [inline test/bidi.spec.mjs:149]: same as above: Why do we need to load/parse the font before each test? If it's not being modified and therefore interfering with following tests, this just adds overhead and increases testing runtime.

File: test/bidi.spec.mjs [hunk 3]
```diff
@@ -133,24 +133,24 @@ describe('bidi.js', function() {
             });
         });
 
-        describe('thai contexts', () => {
+        describe('thai contexts', function () {
             it('should match thai words in a given text', () => {
                 const tokenizer = bidiThai.tokenizer;
-                tokenizer.tokenize('The king said: เป็นคนใจดีสำหรับทุกคน because ความรักคือทุกสิ่ง');
+                bidiThai.processText('The king said: เป็นคนใจดีสำหรับทุกคน because ความรักคือทุกสิ่ง');
```
Comment (Connum) [inline test/bidi.spec.mjs:139]: I'm not sure what that does, can you explain please? Do we need to mix it into this PR?

File: test/glyph.spec.mjs [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
 
-describe('glyph.js', function() {
+describe('glyph.mjs', function() {
 
     describe('lazy loading', function() {
         let font;
         let glyph;
 
-        before(function() {
+        beforeEach(function() {
```
Comment (Connum) [inline test/glyph.spec.mjs:15]: see above

File: test/glyph.spec.mjs [hunk 5]
```diff
@@ -48,7 +48,7 @@ describe('glyph.js', function() {
         let trueTypeFont;
         let openTypeFont;
 
-        before(function() {
+        beforeEach(function() {
```
Comment (Connum) [inline test/glyph.spec.mjs:51]: see above

File: test/glyph.spec.mjs [hunk 6]
```diff
@@ -211,14 +211,14 @@ describe('glyph.js', function() {
...
-describe('glyph.js on low memory mode', function() {
+describe('glyph.mjs on low memory mode', function() {
     let opt = {lowMemory: true};
 
     describe('lazy loading', function() {
         let font;
         let glyph;
 
-        before(function() {
+        beforeEach(function() {
```
Comment (Connum) [inline test/glyph.spec.mjs:221]: see above

File: test/glyph.spec.mjs [hunk 7]
```diff
@@ -254,7 +254,7 @@ describe('glyph.js on low memory mode', function() {
         let trueTypeFont;
         let openTypeFont;
 
-        before(function() {
+        beforeEach(function() {
```
Comment (Connum) [inline test/glyph.spec.mjs:257]: see above

File: src/util.mjs [hunk 8]
```diff
@@ -31,16 +10,6 @@ function arraysEqual(ar1, ar2) {
     return true;
 }
 
-// Check if 2 objects are equal
-function objectsEqual(obj1, obj2) {
```
Comment (Connum) [inline src/util.mjs:35]: I was so sure that we were using this function, but I couldn't actually find it in the code base... strange! This could come in handy for caching transformed glyphs depending on the set variation coordinates, though, and as it's not adding much in file size (and should actually be tree-shaken if I'm not mistaken) and we already have the working test for it - do we really need to remove it?

File: test/util.js [hunk 9]
```diff
@@ -1,18 +0,0 @@
-import assert from 'assert';
-import { objectsEqual } from '../src/util.js';
```
Comment (Connum) [inline test/util.js:2]: see above

**Review by Connum — CHANGES_REQUESTED** [review]
See my comments - other than that, this looks good! Nice work getting rid of all those depedencies and moving towrds a more modern code base.

File: src/util.mjs (hunk 8 above)
Comment (yne) [reply src/util.mjs:35]: I just took the opportunity while doing the cleanup of ther other utils.js dead code  (isBrowser/isNode)

File: test/util.js (hunk 9 above)
Comment (yne) [reply test/util.js:2]: we could ad it back, but it feels weird to add dead code :wink:

File: test/bidi.spec.mjs (hunk 3 above)
Comment (yne) [reply test/bidi.spec.mjs:139]: `tokenizer.tokenize` did not registerContextChecker("thaiWord") so the test failed
`processText` however call tokenizeText() which register it here is the stacktrace

```
  1) bidi.mjs
       thai scripts
         thai contexts
           should match thai words in a given text:
     TypeError: ranges.map is not a function
      at Context.<anonymous> (file:///home/yne/opentype.js/test/bidi.spec.mjs:141:38)
      at process.processImmediate (node:internal/timers:478:21)

```

File: src/font.mjs [hunk 10]
```diff
@@ -628,39 +627,10 @@ Font.prototype.toArrayBuffer = function() {
...
-    } else {
-        const fs = require('fs');
-        const buffer = Buffer.alloc(arrayBuffer.byteLength);
-        const view = new Uint8Array(arrayBuffer);
-        for (let i = 0; i < buffer.length; ++i) {
-            buffer[i] = view[i];
-        }
-        fs.writeFileSync(fileName, buffer);
-    }
+Font.prototype.download = function() {
```
[the remark this answers is not in the document]
Comment (yne) [reply src/font.mjs:632]: It's exact same reason:
all those platform-specific snippet have nothing to do with opentype.js itself and belongs to the documentation (if users can't ask "how to write a node ArrayBuffer to a file" to chatGPT)

I can put the replacement line in the deprecation notice if you want.

```js
fs.writeFileSync("out.otf", Buffer.from(font.toArrayBuffer()));
```
Or the 2 lines browser version:

```js
const href = window.URL.createObjectURL(new Blob([font.toArrayBuffer()]), {type: "font/opentype"});
Object.assign(document.createElement('a'), {download: "out.otf", href}).click();
```

File: src/tables/glyf.mjs [hunk 11]
```diff
@@ -35,18 +35,18 @@ function parseGlyphCoordinate(p, flag, previousValue, shortVectorBitMask, sameBi
 // Parse a TrueType glyph.
 function parseGlyph(glyph, data, start) {
```
[the remark this answers is not in the document]
Comment (yne) [reply src/tables/glyf.mjs:36]: Yes, 2 tests broke as they values are "undefined", here is the stacktrace

```js
  1) glyph.mjs
       lazy loading
         lazily loads numberOfContours:
     AssertionError [ERR_ASSERTION]: undefined == 2
      at Context.<anonymous> (file:///home/yne/opentype.js/test/glyph.spec.mjs:37:20)
      at process.processImmediate (node:internal/timers:478:21)

  2) glyph.mjs on low memory mode
       lazy loading
         lazily loads numberOfContours:
     AssertionError [ERR_ASSERTION]: undefined == 2
      at Context.<anonymous> (file:///home/yne/opentype.js/test/glyph.spec.mjs:243:20)
      at process.processImmediate (node:internal/timers:478:21)
```

Yet the values works for xMin/xMax/yMin/yMax so I simply migrated to the same format as them.

File: src/tables/cpal.mjs [hunk 12]
```diff
@@ -220,7 +220,7 @@ function parseColor(color, targetFormat = 'hexa') {
         if(targetFormat == 'bgra') {
             return color;
         }
-    } else if(globalThis.window && globalThis.window.HTMLCanvasElement && /^[a-z]+$/i.test(color)) {
+    } else if(typeof document !== 'undefined' && /^[a-z]+$/i.test(color)) {
```
[the remark this answers is not in the document]
Comment (yne) [reply src/tables/cpal.mjs:223]: `HTMLCanvasElement` arrived in firefox 1.5 (2005), yet our codebase use much more recent browser API (`Proxy`,`WeakMap`, ... which appeard in Firefox 18 in 2013)

So I think we are pretty much safe for <canvas> usage :)

See [Firefox4 backport](https://github.com/opentypejs/opentype.js/pull/693#issuecomment-2068088620)

File: src/glyphset.mjs [hunk 13]
```diff
@@ -140,6 +140,7 @@ function ttfGlyphLoader(font, index, parseGlyph, data, position, buildPath) {
             return path;
         };
 
+        defineDependentProperty(glyph, 'numberOfContours', '_numberOfContours');
```
[the remark this answers is not in the document]
Comment (yne) [reply src/glyphset.mjs:143]: see below. No Idea if this related to the loss of reify or the bump of Mocha, but it it's related to the PR

File: test/bidi.spec.mjs (hunk 1 above)
Comment (yne) [reply test/bidi.spec.mjs:107]: it allows to use the same codebase on the 3 most popular test runner
yet it has no significant impact on the overall test execution time 

https://mochajs.org/#hooks use `.before()`/`.after()` :cry: 
https://bun.sh/docs/test/lifecycle use `.beforeAll()`/`.afterAll()`
https://jasmine.github.io/api/edge/global use `.beforeAll()`/`.afterAll()`

I could also switch from mocha to `bun test` (and save even more NPM dependencies)
```
npx mocha
npx bun test
npx jasmine "**/*.spec.mjs"
```

Or just create a wrapper in `test/testutil.mjs` with `before = before||beforeAll`

File: test/bidi.spec.mjs (hunk 2 above)
Comment (yne) [reply test/bidi.spec.mjs:149]: There was a bug here
https://github.com/opentypejs/opentype.js/blob/master/test/path.js#L158

where `global.document` were created at init, but deleted after each test (which broke some of them)

**Comment by yne** [comment]
> Nice work getting rid of all those depedencies and moving towrds a more modern code base.

Props to @ILOVEPIE who did the initial migration. I just reused it PR but stopped before doing legacy-compatibility stuff so he could rebase it PR on it

**Comment by ILOVEPIE** [comment]
So, the point of this is to split my pr into two?

File: eslint-local-rules.js
Comment (ILOVEPIE) [inline eslint-local-rules.js:1]: Why was this deleted?

File: package.json [hunk 14]
```diff
@@ -22,76 +22,29 @@
   },
   "main": "./dist/opentype.js",
   "browser": "./dist/opentype.js",
-  "module": "./dist/opentype.module.js",
+  "module": "./dist/opentype.mjs",
   "scripts": {
     "build": "npm run b:umd && npm run b:esm",
     "dist": " npm run d:umd && npm run d:esm",
-    "test": "npm run build && npm run dist && mocha --require reify --recursive && npm run lint",
```
Comment (ILOVEPIE) [inline package.json:29]: I think we still need the ``--recursive``

File: src/opentype.mjs [hunk 15]
```diff
@@ -3,135 +3,46 @@
...
+import meta from './tables/meta.mjs';
+import gasp from './tables/gasp.mjs';
+import svg from './tables/svg.mjs';
+import { PaletteManager } from './palettes.mjs';
 /**
  * The opentype library.
  * @namespace opentype
  */
 
-// File loaders /////////////////////////////////////////////////////////
```
Comment (ILOVEPIE) [inline src/opentype.mjs:47]: I thought we agreed that this would still function in 2.0.0 with a warning, and be removed in a subsequent release.

File: src/opentype.mjs [hunk 16]
```diff
@@ -533,51 +444,20 @@ function parseBuffer(buffer, opt={}) {
  * We use the node.js callback convention so that
  * opentype.js can integrate with frameworks like async.js.
  * @alias opentype.load
- * @param  {string} url - The URL of the font to load.
- * @param  {Function} callback - The callback.
+ * @deprecated
  */
-function load(url, callback, opt = {}) {
```
Comment (ILOVEPIE) [inline src/opentype.mjs:539]: I thought we agreed that this would still function in 2.0.0 with a warning, and be removed in a subsequent release.

File: src/opentype.mjs [hunk 17]
```diff
@@ -533,51 +444,20 @@ function parseBuffer(buffer, opt={}) {
...
 /**
  * Synchronously load the font from a URL or file.
  * When done, returns the font object or throws an error.
  * @alias opentype.loadSync
- * @param  {string} url - The URL of the font to load.
- * @param  {Object} opt - opt.lowMemory
- * @return {opentype.Font}
+ * @deprecated
  */
-function loadSync(url, opt) {
```
Comment (ILOVEPIE) [inline src/opentype.mjs:579]: I thought we agreed that this would still function in 2.0.0 with a warning, and be removed in a subsequent release.

File: test/opentype.spec.mjs [hunk 18]
```diff
@@ -15,33 +15,6 @@ describe('opentype.js', function() {
         assert.equal(aGlyph.path.commands.length, 15);
     });
 
-    it('[deprecated] .load() promise resolve uppon success', function(done) {
```
Comment (ILOVEPIE) [inline test/opentype.spec.mjs:18]: I thought we agreed that this would still function in 2.0.0 with a warning, and be removed in a subsequent release.

File: test/opentype.spec.mjs [hunk 19]
```diff
@@ -84,23 +57,6 @@ describe('opentype.js', function() {
         assert.equal(aGlyph.path.commands.length, 14);
     });
 
-    it('[deprecated] .load() handles a parseBuffer error', function(done) {
```
Comment (ILOVEPIE) [inline test/opentype.spec.mjs:87]: I thought we agreed that this would still function in 2.0.0 with a warning, and be removed in a subsequent release.

File: test/opentype.spec.mjs [hunk 20]
```diff
@@ -119,7 +75,7 @@ describe('opentype.js', function() {
     });
 });
 
-describe('opentype.js on low memory mode', function() {
+describe('opentype.mjs on low memory mode', function() {
```
Comment (ILOVEPIE) [inline test/opentype.spec.mjs:78]: Error here, this should still read opentype.js

**Review by ILOVEPIE — CHANGES_REQUESTED** [review]
I can't approve of this in it's current state.

**Comment by Connum** [comment]
As my brain needed a break from trying to implement CFF2 writing for variable glyphs, I'm currently working on a slim (=really just a couple of lines) plugin API to offer support for standalone CFF1 and PS1/Adobe Type 1 files without bloating the main library.

I think I would be on board with removing the load and download functions directly in 2.0.0 without prior depreciation notice if we were to offer a "browserhelper" and "nodehelper" plugin at the same time, references in a console message when trying to call the functions. The browserhelper could also provide the functionality for adding a preview of the font via CSS @font-face as discussed in #62.

That way, all that users had to do would be to add or import another small file delivered together with our lib in order to keep the functionality.

What do you guys think?

File: eslint-local-rules.js
Comment (yne) [reply eslint-local-rules.js:1]: the usefull rule were put in the package.json eslint rule

File: src/opentype.mjs (hunk 16 above)
Comment (yne) [reply src/opentype.mjs:539]: I thought the function would be *replaced* with a deprecation message

File: test/opentype.spec.mjs (hunk 19 above)
Comment (yne) [reply test/opentype.spec.mjs:87]: the 2.0.0 is a major release
major release does break things
so if we are brewing the 2.0.0 it make more sense to put the warning now to ease this major transition

**Comment by yne** [comment]
> What do you guys think?

for the load/download function, a simple line in the "usage" chapter of the readme would suffice as it is just demonstration of how user can integrate opentype.js in they environment.

but I agree for a separation of concern

This is what jasmine is doing (jasmine/jasmine-core)

File: package.json (hunk 14 above)
Comment (yne) [reply package.json:29]: you are right (226 vs 317)

File: src/font.mjs (hunk 10 above)
[the remark this answers is not in the document]
Comment (yne) [reply src/font.mjs:632]: I've added a corresponding chapter in the README to explain how to get the Font back as file

**Comment by yne** [comment]
> So, the point of this is to split my pr into two?

- Yes this PR remove reify so we can bump Mocha
- Your PR integrate SWC to add a `.legacy.js` build

File: test/util.js (hunk 9 above)
Comment (Connum) [reply test/util.js:2]: ok, let's get rid of it for now...

**Comment by ILOVEPIE** [comment]
Hey, @yne can we get this rebased on the current head so that it can be reviewed and merged if it's ok?

**Comment by yne** [comment]
> Hey, @yne can we get this rebased on the current head so that it can be reviewed and merged if it's ok?

sure, I'll ping you once I get back home (~ end of week)

**Comment by yne** [comment]
@ILOVEPIE It is in fact, already rebased on master.

So there was not much to do

**Review by ILOVEPIE — APPROVED** [review]
Looks good to me.

**Comment by yne** [comment]
We can merge if @Connum is satisfied with the current justifications
